opentelemetry/src/global/trace.rs

```diff
@@ -1,34 +1,177 @@
+    /// regular property. It emphasizes that this operation signifies a
+    /// major change for a `Span` and may lead to re-calculation of sampling or
+    /// filtering decisions made previously depending on the implementation.
+    fn update_name(&mut self, new_name: Cow<'static, str>);
+
+    /// Finishes the `Span`.
+    ///
+    /// Implementations MUST ignore all subsequent calls to `end` (there might be
+    /// exceptions when the tracer is streaming events and has no mutable state
+    /// associated with the Span).
+    ///
+    /// Calls to `end` a Span MUST not have any effects on child `Span`s as they may
+    /// still be running and can be ended later.
+    ///
+    /// This API MUST be non-blocking.
+    fn end(&mut self) {
+        self.end_with_timestamp(crate::time::now());
+    }
+
+    /// Finishes the `Span` with given timestamp
+    ///
+    /// For more details, refer to [`Span::end`]
+    ///
+    /// [`Span::end`]: Span::end()
+    fn end_with_timestamp(&mut self, timestamp: SystemTime);
+}
+
+impl<T: trace::Span> GenericSpan for T {
+    fn add_event_with_timestamp(
+        &mut self,
+        name: Cow<'static, str>,
+        timestamp: SystemTime,
+        attributes: Vec<KeyValue>,
+    ) {
+        self.add_event_with_timestamp(name, timestamp, attributes)
+    }
+
+    fn span_context(&self) -> &SpanContext {
+        self.span_context()
+    }
+
+    fn is_recording(&self) -> bool {
+        self.is_recording()
+    }
+
+    fn set_attribute(&mut self, attribute: KeyValue) {
+        self.set_attribute(attribute)
+    }
+
+    fn set_status(&mut self, code: StatusCode, message: String) {
+        self.set_status(code, message)
+    }
+
+    fn update_name(&mut self, new_name: Cow<'static, str>) {
+        self.update_name(new_name)
+    }
+
+    fn end_with_timestamp(&mut self, timestamp: SystemTime) {
+        self.end_with_timestamp(timestamp)
+    }
+}
+
 /// Wraps the [`BoxedTracer`]'s [`Span`] so it can be used generically by
 /// applications without knowing the underlying type.
 ///
 /// [`Span`]: crate::trace::Span
 #[derive(Debug)]
-pub struct BoxedSpan(Box<DynSpan>);
+pub struct BoxedSpan(Box<dyn GenericSpan + Send + Sync>);
 
-type DynSpan = dyn trace::Span + Send + Sync;
+impl BoxedSpan {
+    pub(crate) fn new<T>(span: T) -> Self
+    where
+        T: GenericSpan + Send + Sync + 'static,
+    {
+        BoxedSpan(Box::new(span))
+    }
+}
 
 impl trace::Span for BoxedSpan {
     /// Records events at a specific time in the context of a given `Span`.
     ///
     /// Note that the OpenTelemetry project documents certain ["standard event names and
     /// keys"](https://github.com/open-telemetry/opentelemetry-specification/tree/v0.5.0/specification/trace/semantic_conventions/README.md)
     /// which have prescribed semantic meanings.
-    fn add_event_with_timestamp(
+    fn add_event_with_timestamp<T>(
         &mut self,
-        name: String,
+        name: T,
```

Comment:
Style nit: I think `impl Into<Cow<'static, str>>` is a little easier to follow (less indirection).